### PR TITLE
[Requirements resolver] Initial integration with the state machine v6

### PR DESCRIPTION
**Pull request by willianrampazzo**
**NOTES**:
- The intent of this PR is to handle the architecture of the requirements resolver. Specifically, the addition of a resolver, changes to the state machine, and changes to the task creation.

This is a minimum set of changes to introduce and integrate the Requirements Resolver with the state machine. The following tests enable the code of this PR (**note**: sudo authority needed).

A single test that passes (`single_success.py`):

```python
from avocado import Test


class PassTest(Test):

    """
    :avocado: requirement={"type": "package", "name": "vim-common"}
    """

    def test(self):
        """
        A test doesn't have to fail to pass
        """
```

Expected output:

```
$ avocado run --test-runner=nrunner single_success.py 
JOB ID     : 9cc29be300f2620006b3290508dbd32127a1877d
JOB LOG    : /home/wrampazz/avocado/job-results/job-2021-04-23T09.27-9cc29be/job.log
 (1/1) single_success.py:PassTest.test: STARTED
 (1/1) single_success.py:PassTest.test: PASS (0.15 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/wrampazz/avocado/job-results/job-2021-04-23T09.27-9cc29be/results.html
JOB TIME   : 7.12 s
```

A single test that fails (`single_fail.py`):

```python
from avocado import Test


class FailTest(Test):

    """
    :avocado: requirement={"type": "package", "name": "vim-common"}
    :avocado: requirement={"type": "package", "name": "bla"}
    :avocado: requirement={"type": "file", "name": "foo.bar"}
    """

    def test(self):
        """
        A fail test
        """
```

Expected output:

```
$ avocado run --test-runner=nrunner single_fail.py 
JOB ID     : 3a24d81bbb1d714461978e45fa6935a89bc14143
JOB LOG    : /home/wrampazz/avocado/job-results/job-2021-04-23T09.27-3a24d81/job.log
RESULTS    : PASS 0 | ERROR 0 | FAIL 0 | SKIP 1 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/wrampazz/avocado/job-results/job-2021-04-23T09.27-3a24d81/results.html
JOB TIME   : 6.17 s
```

**NOTE**: as the test task never starts, it is not accounted for in the Human UI, so it does not show in the execution or in the results.

Multiple tests that pass (`multipletests_success.py`):

```python
from avocado import Test


class PassTest(Test):

    def test_a(self):
        """
        :avocado: requirement={"type": "package", "name": "vim-common"}
        """
    def test_b(self):
        """
        :avocado: requirement={"type": "package", "name": "vim-minimal"}
        """
    def test_c(self):
        """
        :avocado: requirement={"type": "package", "name": "vim-common"}
        """
```

Expected output:

```
$ avocado run --test-runner=nrunner multipletests_success.py 
JOB ID     : 4e2b0a3217838c70a596a125c41ce26387d186fe
JOB LOG    : /home/wrampazz/avocado/job-results/job-2021-04-23T09.28-4e2b0a3/job.log
 (1/3) multipletests_success.py:PassTest.test_a: STARTED
 (1/3) multipletests_success.py:PassTest.test_a: PASS (0.14 s)
 (2/3) multipletests_success.py:PassTest.test_b: STARTED
 (2/3) multipletests_success.py:PassTest.test_b: PASS (0.14 s)
 (3/3) multipletests_success.py:PassTest.test_c: STARTED
 (3/3) multipletests_success.py:PassTest.test_c: PASS (0.14 s)
RESULTS    : PASS 3 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/wrampazz/avocado/job-results/job-2021-04-23T09.28-4e2b0a3/results.html
JOB TIME   : 9.14 s
```

Multiple tests where two fail (`multipletests_fail.py`):

```python
from avocado import Test


class FailTest(Test):

    def test_a(self):
        """
        :avocado: requirement={"type": "package", "name": "vim-common"}
        :avocado: requirement={"type": "package", "name": "bla"}
        """
    def test_b(self):
        """
        :avocado: requirement={"type": "package", "name": "vim-common"}
        """
    def test_c(self):
        """
        :avocado: requirement={"type": "package", "name": "vim-minimal"}
        :avocado: requirement={"type": "package", "name": "bla"}
        """
```

Expected output:

```
$ avocado run --test-runner=nrunner multipletests_fail.py 
JOB ID     : c04c72ebbea91ce45f91cc36c50a5de64c114368
JOB LOG    : /home/wrampazz/avocado/job-results/job-2021-04-23T09.29-c04c72e/job.log
 (2/3) multipletests_fail.py:FailTest.test_b: STARTED
 (2/3) multipletests_fail.py:FailTest.test_b: PASS (0.14 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 2 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/wrampazz/avocado/job-results/job-2021-04-23T09.29-c04c72e/results.html
JOB TIME   : 12.17 s
```

**NOTE**: again, as the test tasks never start, they are not accounted for in the Human UI, so it does not show them in the execution or in the results.

How to play with the requirements package runner:

```
$ avocado-runner-requirement-package runnable-run action=check name=vim-common
{'status': 'started', 'time': 122906.945515359}
{'status': 'running', 'time': 122907.449216436}
{'type': 'stdout', 'log': b'Package vim-common already installed', 'status': 'running', 'time': 122907.4499196}
{'type': 'stderr', 'log': b'', 'status': 'running', 'time': 122907.449974761}
{'result': 'pass', 'status': 'finished', 'time': 122907.450005529}
```

Changes from [v5](https://github.com/avocado-framework/avocado/pull/4572):
- change skip type on tests that should not run locally.
- remove the `type` when passing the dictionary to the runners.
- split the software-manager runner method into multiple action methods.

Changes from [v4](https://github.com/avocado-framework/avocado/pull/4568):
- Improve documentation.
- Adjust the commit message from the commit that removes the checks for `core` requirements.
- Fix the skip message regarding running on CI.
- Skip tests dependent on the package manager to run on CI only.
- Improve the code of `check` on `requirement_package.py` runner.

Changes from [v3](https://github.com/avocado-framework/avocado/pull/4532):
- Use the Avocado Software Manager API instead of the command line in the package runner.
- Docstring adjustments.
- Explicitly fail the package runner when an invalid `action` is passed.
- Adjust `stdout` and `stderr` messages in the package runner.
- Skip spawner check if it was already executed once.
- Split methods on `statemachine` and `runner_nrunner`.
- Remove `core` requirement verification for now.
- Add documentation, examples, and tests.

Changes from [v2](https://github.com/avocado-framework/avocado/pull/4481):
- Change the `avocado-runner-requirement-package` to handle kwargs instead of args.
- Add support to `check` and `remove` packages to the `avocado-runner-requirement-package`.
- Move the requirements resolver back to its own file at `avocado/core/requirements/resolver.py`.
- Add missing support to skip a test if one of its dependencies failed during triage or during the execution.

Changes from [v1](https://github.com/avocado-framework/avocado/pull/4423):
- Requirements are now handled as runners. Each requirement type is mapped to a specific runner. The addition of a new requirement type should be as simple as creating a new runner called `requirement_newtype.py`.
- Moved the resolver to the `avocado/plugins/resolvers.py` file as it is much simpler now.
- Changed the attribute name of the Task class to `dependency` instead of `prereqs`. This maps well to a dependency graph.
- Make the dependencies a set and improve the handling of it in the state machine.

Changes from [v0](https://github.com/avocado-framework/avocado/pull/4369):
- Improve docstring of RequirementRunner;
- Change the Requirements Resolver entry point to a static method;
- Variables renaming to make them more meaningful;
- Improve log message when the requirement type is not available;
- Classes reorder inside the file.

### CODE REVIEW

File: avocado/core/runners/requirement_package.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+                stdout = ''
+                stderr = MESSAGES[cmd]['fail'] % package
+        else:
+            stdout = MESSAGES['check-installed']['fail'] % package
+        return result, stdout, stderr
+
+    def _run_software_manager(self, cmd, package, queue):
+        software_manager = SoftwareManager()
+
+        if not software_manager.is_capable():
```
Comment (clebergnu) [inline avocado/core/runners/requirement_package.py:74]: yay!

File: selftests/unit/test_requirements_resolver.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+                                          {'type': 'package', 'name': 'bar'}])
+        requirements_runnables = RequirementsResolver.resolve(runnable)
+        kind = 'requirement-package'
+        self.assertEqual(kind, requirements_runnables[0].kind)
+        self.assertEqual(kind, requirements_runnables[1].kind)
+        name0 = 'foo'
+        name1 = 'bar'
+        self.assertEqual(name0, requirements_runnables[0].kwargs['name'])
+        self.assertEqual(name1, requirements_runnables[1].kwargs['name'])
+        self.assertIsNone(requirements_runnables[0].kwargs.get('type'))
```
Comment (clebergnu) [inline selftests/unit/test_requirements_resolver.py:22]: yay!

**Review by clebergnu — APPROVED** [review]
LGTM, thanks a million @willianrampazzo !

**Comment by pevogam** [comment]
> Expected output:
> 
> $ avocado run --test-runner=nrunner single_fail.py 
> JOB ID     : 3a24d81bbb1d714461978e45fa6935a89bc14143
> JOB LOG    : /home/wrampazz/avocado/job-results/job-2021-04-23T09.27-3a24d81/job.log
> RESULTS    : PASS 0 | ERROR 0 | FAIL 0 | SKIP 1 | WARN 0 | INTERRUPT 0 | CANCEL 0
> JOB HTML   : /home/wrampazz/avocado/job-results/job-2021-04-23T09.27-3a24d81/results.html
> JOB TIME   : 6.17 s
> 
> NOTE: as the test task never starts, it is not accounted for in the Human UI, so it does not show in the execution or in the results.

Could it instead show with a SKIP rather than PASS line? Is there such a plan in the future?

**Comment by willianrampazzo** [comment]
> > Expected output:
> > $ avocado run --test-runner=nrunner single_fail.py
> > JOB ID     : 3a24d81bbb1d714461978e45fa6935a89bc14143
> > JOB LOG    : /home/wrampazz/avocado/job-results/job-2021-04-23T09.27-3a24d81/job.log
> > RESULTS    : PASS 0 | ERROR 0 | FAIL 0 | SKIP 1 | WARN 0 | INTERRUPT 0 | CANCEL 0
> > JOB HTML   : /home/wrampazz/avocado/job-results/job-2021-04-23T09.27-3a24d81/results.html
> > JOB TIME   : 6.17 s
> > NOTE: as the test task never starts, it is not accounted for in the Human UI, so it does not show in the execution or in the results.
> 
> Could it instead show with a SKIP rather than PASS line? Is there such a plan in the future?

What do you mean @pevogam? Tests that depend on requirements and those are not fulfilled are SKIPed, just like in the output above.

**Comment by pevogam** [comment]
> > > Expected output:
> > > $ avocado run --test-runner=nrunner single_fail.py
> > > JOB ID     : 3a24d81bbb1d714461978e45fa6935a89bc14143
> > > JOB LOG    : /home/wrampazz/avocado/job-results/job-2021-04-23T09.27-3a24d81/job.log
> > > RESULTS    : PASS 0 | ERROR 0 | FAIL 0 | SKIP 1 | WARN 0 | INTERRUPT 0 | CANCEL 0
> > > JOB HTML   : /home/wrampazz/avocado/job-results/job-2021-04-23T09.27-3a24d81/results.html
> > > JOB TIME   : 6.17 s
> > > NOTE: as the test task never starts, it is not accounted for in the Human UI, so it does not show in the execution or in the results.
> > 
> > 
> > Could it instead show with a SKIP rather than PASS line? Is there such a plan in the future?
> 
> What do you mean @pevogam? Tests that depend on requirements and those are not fulfilled are SKIPed, just like in the output above.

Yes, I mostly meant printing the SKIP line (without a STARTED line) in the UI because the current output makes it harder to see which tests were skipped. But I guess this might also be harder to implement and less compatible with current architecture decisions.

**Comment by willianrampazzo** [comment]
> > > > Expected output:
> > > > $ avocado run --test-runner=nrunner single_fail.py
> > > > JOB ID     : 3a24d81bbb1d714461978e45fa6935a89bc14143
> > > > JOB LOG    : /home/wrampazz/avocado/job-results/job-2021-04-23T09.27-3a24d81/job.log
> > > > RESULTS    : PASS 0 | ERROR 0 | FAIL 0 | SKIP 1 | WARN 0 | INTERRUPT 0 | CANCEL 0
> > > > JOB HTML   : /home/wrampazz/avocado/job-results/job-2021-04-23T09.27-3a24d81/results.html
> > > > JOB TIME   : 6.17 s
> > > > NOTE: as the test task never starts, it is not accounted for in the Human UI, so it does not show in the execution or in the results.
> > > 
> > > 
> > > Could it instead show with a SKIP rather than PASS line? Is there such a plan in the future?
> > 
> > 
> > What do you mean @pevogam? Tests that depend on requirements and those are not fulfilled are SKIPed, just like in the output above.
> 
> Yes, I mostly meant printing the SKIP line (without a STARTED line) in the UI because the current output makes it harder to see which tests were skipped. But I guess this might also be harder to implement and less compatible with current architecture decisions.

Ah, okay, I got it, and I agree it makes sense. I'll keep that on the upcoming plans.

Right now it does not do that because the test is never started. Just tests that are started are marked with a SKIP line when they are skipped. That is something we need to change.

**Pull request merged**